osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs

```diff
@@ -41,13 +49,21 @@ public class DifficultyHitObject
         /// <param name="hitObject">The <see cref="HitObject"/> which this <see cref="DifficultyHitObject"/> wraps.</param>
         /// <param name="lastObject">The last <see cref="HitObject"/> which occurs before <paramref name="hitObject"/> in the beatmap.</param>
         /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
-        public DifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate)
+        /// <param name="objects">The list of <see cref="DifficultyHitObject"/>s in the current beatmap.</param>
+        /// <param name="position">The position of this <see cref="DifficultyHitObject"/> in the <see cref="difficultyHitObjects"/> list.</param>
+        public DifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate, List<DifficultyHitObject> objects, int position)
         {
+            difficultyHitObjects = objects;
+            Position = position;
```

Comment:
I can't really find myself agreeing with this change. Relying on the object count when constructing difficulty hitobject locks the users into having to construct the object sequentially, but the API does not communicate clearly that this is the case. I preferred explicit index passing.

Reply:
I think I'd prefer making it clear that the method should generate difficulty hitobjects sequentially, unless I'm missing some specific use cases which would be entirely reasonable - open to more opinions on this

Reply:
I get where @bdach is coming from, short of that `DifficultyHitObject` is now computing its position by guessing that it will only ever be added to the end of the list, and it's not documented anywhere that that can be the case.

I would be fine with bringing back the constructor parameter but still pass `list.Count` in usages rather than trying to compute it from the for-loop index.

I'm also gonna note that you can alternatively avoid the necessity of a `Position` by computing the index of the object directly instead (`list.IndexOf(this)` during `Previous`/`Next` calls), but that might be unnecessarily expensive, not sure.

Reply:
Done as suggested - will keep the `Position` property around though since it has use across various parts of difficulty calculation